test/xds/xds_client_outlier_detection_test.go

```diff
@@ -0,0 +1,218 @@
+	}
+}
+
+func defaultClusterWithOutlierDetection(clusterName, edsServiceName string, secLevel e2e.SecurityLevel) *v3clusterpb.Cluster {
+	cluster := e2e.DefaultCluster(clusterName, edsServiceName, secLevel)
+	cluster.OutlierDetection = &v3clusterpb.OutlierDetection{
+		Interval:                       &durationpb.Duration{Nanos: 50000000}, // .5 seconds
+		BaseEjectionTime:               &durationpb.Duration{Seconds: 30},
+		MaxEjectionTime:                &durationpb.Duration{Seconds: 300},
+		MaxEjectionPercent:             &wrapperspb.UInt32Value{Value: 1},
+		FailurePercentageThreshold:     &wrapperspb.UInt32Value{Value: 50},
+		EnforcingFailurePercentage:     &wrapperspb.UInt32Value{Value: 100},
+		FailurePercentageRequestVolume: &wrapperspb.UInt32Value{Value: 1},
+		FailurePercentageMinimumHosts:  &wrapperspb.UInt32Value{Value: 1},
+	}
+	return cluster
+}
+
+// TestOutlierDetectionWithOutlier tests the Outlier Detection Balancer e2e. It
+// spins up three backends, one which consistently errors, and configures the
+// ClientConn using xDS to connect to all three of those backends. The Outlier
+// Detection Balancer should eject the connection to the backend which
+// constantly errors, and thus RPC's should mainly go to backend 1 and 2.
+func (s) TestOutlierDetectionWithOutlier(t *testing.T) {
+	oldOD := envconfig.XDSOutlierDetection
+	envconfig.XDSOutlierDetection = true
+	internal.RegisterOutlierDetectionBalancerForTesting()
+	defer func() {
+		envconfig.XDSOutlierDetection = oldOD
+		internal.UnregisterOutlierDetectionBalancerForTesting()
+	}()
+
+	managementServer, nodeID, _, resolver, cleanup := e2e.SetupManagementServer(t, nil)
+	defer cleanup()
+
+	// counters for how many times backends got called
+	var count1, count2, count3 int
+
+	// Working backend 1.
+	port1, cleanup1 := startTestService(t, &stubserver.StubServer{
+		EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) {
+			count1++
+			return &testpb.Empty{}, nil
+		},
+		Address: "localhost:0",
+	})
+	defer cleanup1()
+
+	// Working backend 2.
+	port2, cleanup2 := startTestService(t, &stubserver.StubServer{
+		EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) {
+			count2++
+			return &testpb.Empty{}, nil
+		},
+		Address: "localhost:0",
+	})
+	defer cleanup2()
+	// Backend 3 that will always return an error and eventually ejected.
+	port3, cleanup3 := startTestService(t, &stubserver.StubServer{
+		EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) {
+			count3++
+			return nil, errors.New("some error")
+		},
+		Address: "localhost:0",
+	})
+	defer cleanup3()
+
+	const serviceName = "my-service-client-side-xds"
+	resources := defaultClientResourcesMultipleBackendsAndOD(e2e.ResourceParams{
+		DialTarget: serviceName,
+		NodeID:     nodeID,
+		Host:       "localhost",
+		SecLevel:   e2e.SecurityLevelNone,
+	}, []uint32{port1, port2, port3})
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*100)
```

Comment:
Prefer using the already available default test timeout const. `defaultTestShortTimeout`.
`100s` is too large anyways.